### PR TITLE
initial commit of ajax image gallery

### PR DESCRIPTION
**Pull request by ghost**
Ok, so these are the first drafts. There are still a lot of things to be done. Any idea how to stop the page from being automatically refreshed after image upload? This issue is not happening on my own CMS.

Don't mind the display_errors in index.php. I will remove them on next commit.

### CODE REVIEW

File: module/LearnZF2Ajax/test/PHPUnit/LearnZF2AjaxTest/Controller/IndexControllerTest.php [hunk 1]
```diff
@@ -1,79 +1,79 @@
...
-        $this->dispatch('/learn-zf2-ajax', 'POST', $postData, true);
-        $request = $this->getRequest();
-
-        $this->assertTrue($request->isXmlHttpRequest());
-        $this->assertResponseHeaderContains('Content-Type', 'application/json; charset=utf-8');
-
-        $this->assertEquals('{"form":{"username":{},"password":{}},"data":{"result":true,"message":"Ajax request success"}}', $this->getResponse()->getBody());
-    }
-}
+<?php
```
Comment (samsonasik) [inline module/LearnZF2Ajax/test/PHPUnit/LearnZF2AjaxTest/Controller/IndexControllerTest.php]: why you remove/cut LearnZF2Ajax test ? it should be copy if you want as reference

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    /**
+     * Upload all images async
+     *
+     * @param  array $files
+     * @return Response containing headers with information about each image
+     */
+    private function uploadImages()
+    {
+        $this->view->setTerminal(true);
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php]: iirc, if we are using <code>acceptableviewmodelselector</code> plugin, we don't need setTerminal() as far as the <code>ContentType</code> and <code>Accept</code> header is <code>application/json</code>

File: public/js/javascript.js [hunk 3]
```diff
@@ -1,24 +1,141 @@
...
+             */
+            $("#imgajax").change(function (e) {
+                var $form = $("#content").submit();
+                $.ajax({
+                    url: $form.attr("action"),
+                    type: "POST",
+                    data: new FormData($form[0]),
+                    processData:false,
+                    contentType:  false,
+                    cache: false,
```
Comment (samsonasik) [inline public/js/javascript.js]: we need <code>accepts: "application/json"</code> to be handled by <code>acceptableviewmodelselector()</code> plugin in controller.

File: module/LearnZF2AjaxImageGallery/test/PHPUnit/phpunit.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+                <directory suffix=".php">./../</directory>
+            </exclude>
+        </whitelist>
+        <blacklist>
+            <directory suffix=".php">./../../../../vendor/</directory>
+        </blacklist>
+    </filter>
+    <logging>
+        <log type="coverage-html" target="./../../../../data/test-html-report/learnzf2ajax/" charset="UTF-8" yui="true" />
+        <log type="coverage-php" target="./../../../../build/coverage/coverage-learnzf2ajax.cov"/>
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/test/PHPUnit/phpunit.xml]: change : s/learnzf2ajax/learnzf2ajaximageagallery

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                }
+
+                if ($adapter->receive($file["name"])) {
+                    $uploadStatus["successFiles"][] = $file["name"]." was uploaded successfully";
+                } else {
+                    $uploadStatus["errorFiles"][] = $file["name"]. " was not uploaded";
+                }
+            }
+            echo Json::encode($uploadStatus);
+            exit;
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php]: this should be return new JsonModel() if we send <code>Accept: application/json</code> in request.

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    protected $acceptCriteria = [
+        'Zend\View\Model\JsonModel' => ['application/json'],
+        'Zend\View\Model\ViewModel' => ['text/html'],
+    ];
+
+    public $view;
+
+    public function __construct()
+    {
+        $this->view = new ViewModel();
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php]: we need to rely on <code>acceptableviewmodelselector</code> to choose what kind of ViewModel will be used based on accept header.

File: public/js/javascript.js [hunk 7]
```diff
@@ -1,24 +1,141 @@
...
+        },
+
+        /**
+         * Gallery view
+         */
+        showFiles: function () {
+                $(".uploader-inline, .large-image").hide();
+                $(".gallery-view").find("figure.centered").not(".large-image").remove();
+                $(".gallery-view, .ajax-loader").show();
+            $.get( "/learn-zf2-ajax-image-gallery/index/files", function (files) {
```
Comment (samsonasik) [inline public/js/javascript.js]: we should avoid using "/" directly, you can add : 
```php
<script type="text/javascript">
var basePath = '<?php echo $this->basePath(); ?>';
</script>
```
in <code>Application/view/layout/header.phtml</code> before <code>headScript</code> view helper called, and after that, you can use : 
```javascript
$.get(basePath +  "/learn-zf2-ajax-image-gallery/index/files", function (files) {
```

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+        $this->add([
+            'name' => 'imageUpload',
+            'type'  => 'File',
+            'attributes' => [
+                'class' => 'imgupload',
+                'id' => 'imgajax',
+                'multiple' => true,
+            ],
+        ]);
+    }
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php]: we need to use <code>FormElementManager</code> to get the advantage of ServiceManager. define in <code>init()</code> and <code>getInputFilterSpecification</code>.

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        $this->createDirectories();
+        $adapter->setValidators([$size, $isImage, $extension]);
+        $adapter->setDestination('public/userfiles/'.date("Y_M").'/images/');
+        $this->upload($adapter);
+    }
+
+    /**
+     * @param  Zend\File\Transfer\Adapter\Http $adapter
+     */
+    private function upload($adapter)
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php]: I think we can replace usage of <code>Zend\File\Transfer\Adapter\Http</code>, with define filters and validators in form::getInputFilterSpecification()

**Comment by samsonasik** [comment]
I don't have idea about : 
> refreshed after image upload

it should be because the javascript is break somewhere, should can be check with firebug

**Comment by Stanimirdim92** [comment]
Thanks for the feedback. I learned few new things :). Will fix everything today or tomorrow. 
As for the test cases, I have no idea why they were deleted from the Ajax folder, since I still have them on my localhost, but ignore them for now.

**Comment by samsonasik** [comment]
:+1: , anyway, please run  
```
php public/index.php migrations:generate
```
and add new sql in generated file like this : 
```php
<?php

namespace DoctrineORMModule\Migrations;

use Doctrine\DBAL\Migrations\AbstractMigration;
use Doctrine\DBAL\Schema\Schema;

/**
 * Auto-generated Migration: Please modify to your needs!
 */
class Version20150606222937 extends AbstractMigration
{
    /**
     * @param Schema $schema
     */
    public function up(Schema $schema)
    {
        // this up() migration is auto-generated, please modify it to your needs
        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
        $this->addSql("
            INSERT INTO `module_list` (`module_name`, `module_desc`, `module_route`) VALUES
            ('LearnZF2AjaxImageGallery', 'Learn Ajax Image Gallery with ZF2', 'learn-zf2-ajax-image-gallery');");
    }

    /**
     * @param Schema $schema
     */
    public function down(Schema $schema)
    {
        // this down() migration is auto-generated, please modify it to your needs
        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
        $this->addSql("
            DELETE FROM `module_list` where `module_name` = 'LearnZF2AjaxImageGallery'");
    }
}
```

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php (hunk 8 above)
Comment (Stanimirdim92) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php]: I have a xModel which implements InputFilterAwareInterface with get|setInputFIlter() and I have a form which implements InputFilterProviderInterface with init() and getInputFilterSpecification() functions. Both input filters validates the form according to their interface functions. Which one should I use? Should I leave them like that, or I can move the getInputFilterSpecification into the xModel class.

Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php]: as we will use FormElementManager, new getInputFilter() |setInputFilter() methods are not needed in both Model and Form. getInputFilter() is callable by form instance

Comment (Stanimirdim92) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Form/AjaxImageUploadForm.php]: Thanks! Now I can have even bigger cohesion in my models.

**Comment by samsonasik** [comment]
please add  `.gitignore` to `public/userfiles/2015_Jun/images/` or `public/userfiles` with `.gitignore` content: 
```
*
!.gitignore
```
to avoid commited local images upload files.

Thank you ;)

**Comment by samsonasik** [comment]
please rebase with latest upstream master with the following steps:
```bash
$ git remote add upstream https://github.com/sitrunlab/LearnZF2.git
$ git checkout master && git pull upstream master
$ git checkout changelog && git rebase master
```
you will get some conflict, you need to clean up : `<<<<HEAD` conflict. After it clean, Don't add a commit, but you can do: 
```bash
$ rm -rf composer.lock && rm -rf vendor/ && composer install
$ git add .
$ git rebase --continue
```
Now, you need to force update origin:
```bash
$ git push --force origin changelog
```

**Comment by Stanimirdim92** [comment]
I have commited a new code. There are still some bugs to be fixes before I write the unit tests.

**Comment by samsonasik** [comment]
please do rebase instead of merge like i noted above.

**Comment by samsonasik** [comment]
@Stanimirdim92 I can see that there is no conflict with master branch now. You can still use this branch. For travis failure, it should will be rebuilt on your next commit. Thanks ;)

**Comment by samsonasik** [comment]
your composer.lock conflict now with latest master, please try rebase ;)

**Comment by Stanimirdim92** [comment]
Hi,

Can someone of you guys have a look at uploadFiles() function? It's inside the controller. I spent like half a day trying to figure out why on earth it doesn't work with return JsonModel([]) nor if I use return Json::encode($uploadStatus), while every other method works, but for this, I couldn't fix the problem.

I'm also working on a image processing library based on GD extension. So far so good, just needs to polishing and filter effects manipulation. After that I will update this PR.

Regards,
Stanimir

File: module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                    }
+                } else {
+                    foreach ($adapter->getMessages() as $key => $msg) {
+                        $uploadStatus["errorFiles"][] = $file["name"]." ".strtolower($msg);
+                    }
+                }
+            }
+        }
+        // JsonModel doesn't work... It returns the page html even file upload s successful
+        echo Json::encode($uploadStatus);
```
Comment (samsonasik) [inline module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php]: you need to set in javascript: 
```javascript
"Accept": "application/json"
```
when call xmlHttpRequest to make `JsonModel` works

**Comment by Stanimirdim92** [comment]
Can you confirm that all dependencies were resolved? The main problem was that my composer.lock file from changelog branch was way ahead of the one from master and everytime when I try to rebase it would conflict and ask me to merge or rebase won't be started. For now everything seems to be up to date.

**Comment by samsonasik** [comment]
It seems your branch already overlapped. I think you can focus on finishing it now, I will take care the composer.lock issue and cs fixes when it ready for merge. Please see my comment before about doctrine migration, and unit tests. You can run unit tests locally so when the unit tests ready, you can ping us ;)

**Comment by Stanimirdim92** [comment]
Needs review. @samsonasik  @acelaya

**Comment by samsonasik** [comment]
Thanks, I will try to do it today your time ;)

**Comment by samsonasik** [comment]
I pulled your branch.
1. it seems you didn't run `php public/index.php migrations:generate` and add new sql as I noted above https://github.com/sitrunlab/LearnZF2/pull/136#issuecomment-109604431
2. I tried upload and got internal server error? any php configuration needed?

**Comment by Stanimirdim92** [comment]
I executed the command and commited the modified file. No 500 errors. What's the log saying? Skype is down, but I'm at facebook ad zftalk channel

**Comment by samsonasik** [comment]
I debugged it and found that directory target doesn't exists, added .gitignore locally for that, and fixes migration issue locally ( redeclare class exists).

**Comment by samsonasik** [comment]
@Stanimirdim92  merged, thanks!, Next is make unit tests coverage 100% ;), Currently I see still 93.33% for this module.
